app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.service
+
+import uk.gov.hmrc.domain.Nino
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext._
+import uk.gov.hmrc.tai.model.TaxYear
+import uk.gov.hmrc.tai.util.constants.journeyCache.UpdateNextYearsIncomeConstants
+
+import scala.concurrent.Future
+
+final case class UpdateNextYearsIncomeCacheModel(employmentName: String, employmentId: Int, currentValue: BigDecimal, newValue: Option[Int] = None) {
+  def toCacheMap: Map[String, String] = {
+    Map(
+      UpdateNextYearsIncomeConstants.EMPLOYMENT_NAME -> employmentName,
+      UpdateNextYearsIncomeConstants.EMPLOYMENT_ID -> employmentId.toString,
+      UpdateNextYearsIncomeConstants.CURRENT_AMOUNT -> currentValue.toString
+    ) ++ { newValue match {
+      case Some(_) => Map(UpdateNextYearsIncomeConstants.NEW_AMOUNT -> newValue.get.toString)
+      case None => Map()
+    }}
+  }
+}
+
+class UpdateNextYearsIncomeService {
+
+  lazy val journeyCacheService: JourneyCacheService = JourneyCacheService(UpdateNextYearsIncomeConstants.JOURNEY_KEY)
+  lazy val employmentService: EmploymentService = EmploymentService
+  lazy val taxAccountService: TaxAccountService = TaxAccountService
+
+  def setup(employmentId: Int, nino: Nino)(implicit hc: HeaderCarrier): Future[UpdateNextYearsIncomeCacheModel] = {
+    val taxCodeIncomeFuture = taxAccountService.taxCodeIncomeForEmployment(nino, TaxYear(), employmentId)
+    val employmentFuture = employmentService.employment(nino, employmentId)
+
+    for {
+      taxCodeIncomeOption <- taxCodeIncomeFuture
+      employmentOption <- employmentFuture
+    } yield (taxCodeIncomeOption, employmentOption) match {
+      case (Some(taxCodeIncome), Some(employment)) => {
+        val model = UpdateNextYearsIncomeCacheModel(employment.name, employmentId, taxCodeIncome.amount)
+
+        journeyCacheService.cache(model.toCacheMap)
+
+        model
+      }
+      case _ => throw new RuntimeException("[UpdateNextYearsIncomeService] Could not set up next years estimated income journey")
```

Comment:
I'm not a big fan of this throwing error way of handling cases where there is no data. We can probably discuss with design on if we can show a more meaningful page than the standard error page

Reply:
This is true, we dont want to block the rest of the work based on this though. Perhaps a TODO would be appropriate